bundles/org.openhab.binding.venstarthermostat/src/main/java/org/openhab/binding/venstarthermostat/internal/handler/VenstarThermostatHandler.java

```diff
@@ -318,23 +336,93 @@ private VenstarSystemMode getSystemMode() {
+        params.put("away", "" + away.mode());
+        String updateURLextension = "settings";
+        updateThermostat(updateURLextension, params);
+    }
+
+    private void updateControls(int heat, int cool, VenstarSystemMode mode) {
+        // this function corresponds to the thermostat local API POST /control instruction
+        // the function can be expanded with other parameters which are changed via POST /control
         Map<String, String> params = new HashMap<>();
-        log.debug("Updating thermostat {}  heat:{} cool {} mode: {}", getThing().getLabel(), heat, cool, mode);
         if (heat > 0) {
             params.put("heattemp", String.valueOf(heat));
         }
         if (cool > 0) {
             params.put("cooltemp", String.valueOf(cool));
         }
         params.put("mode", "" + mode.mode());
+        String updateURLextension = "control";
+        updateThermostat(updateURLextension, params);
+    }
+
+    private void updateThermostat(String updateType, Map<String, String> params) {
+        // this function is called from either updateControl or updateSettings and sends the command to the thermostat
+        // add other API variables here if they get programmed in future
+
+        // settings
+        VenstarAwayMode away = infoData.getAway();
+
+        // controls
+        double heat = infoData.getHeattemp();
+        double cool = infoData.getCooltemp();
+        VenstarSystemMode mode = infoData.getMode();
+
+        if (updateType == "settings") {
+            if (params.containsKey("away")) {
+                String awayString = params.get("away");
+                if (awayString != null) {
+                    int awayInt = Integer.parseInt(awayString);
+                    away = VenstarAwayMode.fromInt(awayInt);
+                }
+
+            }
+
+            // other thermostat settings can be added here in the same way
+            log.debug("Updating thermostat {} with settings {}", getThing().getLabel(), params);
+        }
+        if (updateType == "control") {
+            if (params.containsKey("heattemp")) {
+                String heatString = params.get("heattemp");
+                if (heatString != null) {
+                    heat = Double.parseDouble(heatString);
+                }
+            }
+            if (params.containsKey("cooltemp")) {
+                String coolString = params.get("cooltemp");
+                if (coolString != null) {
+                    cool = Double.parseDouble(coolString);
+                }
+            }
+            if (params.containsKey("mode")) {
+                String modeString = params.get("mode");
+                if (modeString != null) {
+                    int modeInt = Integer.parseInt(modeString);
+                    mode = VenstarSystemMode.fromInt(modeInt);
+                }
+
+            }
+            // other thermostat controls can be added here in the same way
+        }
         try {
```

Comment:
This should probably be the only logic in this function, actually sending the data to the thermostat and getting the result.

Reply:
OK. I will take a look at it again tonight (busy during the day, unfortunately). I found it a bit complicated to correctly pass and convert the data types of the Map<String,String> between the functions - the native variables being passed are not Strings (e.g. int heat) so there are a few conversions going on. Plus, the Map "get" function can return null, which leads to some additional code to make sure it's non-null. I'm pretty sure it could all be cleaned up, because in the end the thermostat just needs a String for the URL, but it would take a bit of effort on several classes in the binding. What do you think?

Reply:
@digitaldan I changed the code again. It is quite difficult to call updateThermostat from updateSettings and updateControls because the "try" block in updateThermostat includes an update to the infoData object which requires the heat, cool, mode, awaymode objects (not String equivalents which are passed in the Map<>). I therefore decided the simplest way is to include the try/catch block in both updateSettings and updateControls functions and delete the updateThermostat function altogether. Interested if you can see an easy way to do it differently as you suggested above.